### ISSUE TITLE
Feature request: Fragment support

### ISSUE THREAD

**Issue by bannmann** [tomokinakamaru/silverchain#66]
I have several APIs where I need to repeat the exact same expression in several places. Take this (simplified) example from [PureTemplate](https://github.com/puretemplate/puretemplate):

```java
GroupLoader
{
    Group

    // Allow only one call of each with*() method, but in any order, and mixed with other, unrestricted methods
    ( importTemplates() | registerModelAdaptor() | registerAttributeRenderer() )*
    {
        (
            withDelimiters()?
            ( importTemplates() | registerModelAdaptor() | registerAttributeRenderer() )*
        ),
        (
            withErrorListener()?
            ( importTemplates() | registerModelAdaptor() | registerAttributeRenderer() )*
        ),
        (
            withLegacyRendering()?
            ( importTemplates() | registerModelAdaptor() | registerAttributeRenderer() )*
        )
    }

    build();
}
```

It's plain to see that the lines starting with `( importTemplates() ...` are identical. Note that in the [original version](https://github.com/puretemplate/puretemplate/blob/db889ae0f9678482ec739dce887a458720fe93e2/src/main/silverchain/puretemplate.ag#L71) , those lines are much longer (even with the recent addition of import declarations).

For this and other situations, Silverchain could offer a feature that lets me define an alias for an expression that I can reuse throughout my AG file. I think "fragment" might be a good name. I'm not sure what the syntax should be - as usual we should try to balance "being obvious", i.e. Java-like, with achieving compact syntax.

Here's one idea for defining and using such a fragment:

```java

$UNRESTRICTED_OPTIONS = ( importTemplates() | registerModelAdaptor() | registerAttributeRenderer() )*;

GroupLoader
{
    Group

    // Allow only one call of each with*() method, but in any order, and mixed with other, unrestricted methods
    $UNRESTRICTED_OPTIONS
    {
        ( withDelimiters()? $UNRESTRICTED_OPTIONS ),
        ( withErrorListener()? $UNRESTRICTED_OPTIONS ),
        ( withLegacyRendering()? $UNRESTRICTED_OPTIONS )
    }

    build();
}
```

This idea imitates Java field declaration and definition, uses ALL_CAPS (which may or may not be enforced by Silverchain) like Java constants and prefixes the name with a `$` (which should be enforced) to reduce the chance of collisions with class/method/parameter names.

Alternatively, one might use existing keywords as part of the syntax. I came up with `private static final $UNRESTRICTED_OPTIONS = ...`,  intentionally mimicking a Java constant, but that feels "fake" to me because none of the three keywords _really_ do what they do in Java here - and it's really long-winded.

Lastly, we could invent new keywords, something like `fragment $UNRESTRICTED_OPTIONS = ...`. One could say that version is _bad_ because it's not Java-like, or _great_ because the keyword makes its purpose obvious. I have a slight preference for the latter. However, for consistency, we should then definitely consider having keywords elsewhere in AG, for example `class GroupLoader { }` and `rule MyReturnValue Method1() Method2();`.

Note: implementation-wise, any usage of a fragment would be equivalent of pasting the expression at that position, but wrapped in parentheses to make sure the fragment is atomic. This would even allow things like `$UNRESTRICTED_OPTIONS*`.

Looking forward to read your thoughts on this!

**Comment by tomokinakamaru** [comment]
Thank you for the request :)

## Notation

I prefer the first notation `$FOO = ...;` without enforcing ALL_CAPS. I would also give ALL_CAPS names to fragments, but it can be a convention like in Java.

> but that feels "fake" to me because none of the three keywords really do what they do in Java here - and it's really long-winded.

Agree, I also feel redundant because Silverchain does not provide `public` or non-`final` fragments.

> However, for consistency, we should then definitely consider having keywords elsewhere in AG, for example `class GroupLoader { }` and `rule MyReturnValue Method1() Method2();`.

Adding keywords elsewhere definitely makes every part easy to understand, so we should work on it. However, we first need to discuss the grammar of AG in another issue; There are many points that we need to fix, since the current grammar is based on what I have designed long time ago before Silverchain focuses on the Java API generation :(

## Fragment in Fragment

To reduce the repetition of an expression as much as possible, I think Silverchain should allow its users to put `$FRAGMENT` in the right hand side of a fragment statement (e.g. `$FOO = method() $BAR;`).

I am planning to implement the _imperative_ interpretation of fragment statements, which **does not** allow forward references:
```
$FOOS_AND_BARS = foo()+ $BARS ; // <- Forward reference, error
$BARS = bar()+ ;
```
This way of interpretation is the same as the one used in the field declaration in Java.

For consistency with the class declaration, we can alternatively implement the _declarative_ interpretation, which **does** allow forward references:
```
$FOOS_AND_BARS = foo()+ $BARS ; // <- No problem
$BARS = bar()+ ;
```
This way is less restrictive, but I think it's freedom that Silverchain users do not need. Furthermore, Silverchain needs to check the dependencies among fragments before code generation, to avoid infinite loop of fragment expansion. (Silverchain should throw an error before expanding `$FOO` and `$BAR` in `$FOO = method()+ $BAR; $BAR = method()+ $FOO;`.)

To summarize, I would like to hear your opinion on the following:

- Should Silverchain support fragment-in-fragment?

- Should Silverchain interpret fragment statements imperatively?

(My answer is YES for both)

**Comment by bannmann** [comment]
> Should Silverchain support fragment-in-fragment?

Sure! Come to think of it, I think it's both useful and will simply be expected by developers.

> Should Silverchain interpret fragment statements imperatively?

Yes. I see no need for the freedom granted by declarative interpretation either.

**Comment by bannmann** [comment]
> However, we first need to discuss the grammar of AG in another issue

I'd love to do that!

Depending on the type of discussion, we could also take it to https://gitter.im/ which may be more suited for back-and-forth style conversations. However, so far our _explain - ask questions - get opinions_ loop seems to work fine here :-)

**Comment by bannmann** [comment]
@tomokinakamaru, I found your commit 9e06d00951869643accd7b191131b5ff7066bab1 on `feature/fragment` and took the liberty of testing it with my code. The fragments feature works flawlessly; I was able to eliminate all redundancies.

**Comment by tomokinakamaru** [comment]
@bannmann Sorry for my late reply, it took long to update documents (#67) because I am not good at writing documents :(

Thank you for testing the fragment feature! I was going to ask you :) I will merge the branch `feature/fragment`.

---

[Gitter]
We can move to Gitter! It seems nice for back-and-forth style conversations though I have never used Gitter before. (If Gitter does not fit, we can move back here.) Anyway, I will create a community on Gitter later :)
